python-checks/src/main/java/org/sonar/python/checks/SelfAssignmentCheck.java

```diff
@@ -40,72 +47,82 @@
   private Set<String> importedNames = new HashSet<>();
 
   @Override
-  public void visitFile(AstNode node) {
-    importedNames.clear();
-  }
-
-  @Override
-  public Set<AstNodeType> subscribedKinds() {
-    return immutableSet(
-      PythonGrammar.EXPRESSION_STMT,
-      PythonGrammar.IMPORT_NAME,
-      PythonGrammar.IMPORT_AS_NAME);
-  }
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.FILE_INPUT, ctx -> this.importedNames.clear());
 
-  @Override
-  public void visitNode(AstNode node) {
-    if (node.is(PythonGrammar.IMPORT_NAME)) {
-      for (AstNode dottedAsName : node.select().children(PythonGrammar.DOTTED_AS_NAMES).children(PythonGrammar.DOTTED_AS_NAME)) {
-        AstNode importedName = dottedAsName.getFirstChild().getLastChild(PythonGrammar.NAME);
-        addImportedName(dottedAsName, importedName);
+    context.registerSyntaxNodeConsumer(Tree.Kind.IMPORT_FROM, ctx -> {
+      PyImportFromTree importFrom = (PyImportFromTree) ctx.syntaxNode();
```

Comment:
Nitpick : in order to keep it in the lambda expression I believe a one liner would be clearer here. 

```
( (PyImportFromTree) ctx.syntaxNode()).importedNames().stream().forEach(this::addImportedName)
```